File: apps/web/shared/liveBrowser.ts
import { z } from "zod";
import { browserSkillIdSchema } from "./browserSkills";

export const liveBrowserActorTypeValues = [
  "agent",
  "user",
  "system",
  "policy",
] as const;

export const liveBrowserSourceTypeValues = [
  "automation",
  "chat",
  "workflow",
  "agency",
] as const;

export const liveBrowserSessionStatusValues = [
  "created",
  "provisioning",
  "ready",
  "agent_running",
  "waiting_for_human",
  "human_controlling",
  "waiting_for_runtime_recovery",
  "failed_recovery_required",
  "completed",
  "cancelled",
  "failed",
  "expired",
] as const;

export const liveBrowserControlModeValues = [
  "observe",
  "approve_only",
  "takeover",
  "agent_control",
] as const;

export const liveBrowserAssistRequestTypeValues = [
  "decision",
  "field_input",
  "review_page",
  "takeover_required",
] as const;

export const liveBrowserAssistRequestStatusValues = [
  "pending",
  "resolved",
  "cancelled",
] as const;

export const liveBrowserEventTypeValues = [
  "session_created",
  "session_state_changed",
  "stream_ready",
  "frame_updated",
  "url_changed",
  "command_queued",
  "command_started",
  "command_completed",
  "command_failed",
  "assist_requested",
  "assist_resolved",
  "approval_requested",
  "approval_resolved",
  "takeover_started",
  "takeover_lease_expiring",
  "takeover_ended",
  "incident",
  "agent_started",
  "agent_resumed",
  "navigation_completed",
  "session_completed",
  "session_failed",
] as const;

export const liveBrowserErrorCodeValues = [
  "session_version_conflict",
  "session_not_found",
  "session_terminated",
  "invalid_state_transition",
  "policy_denied",
  "rate_limited",
  "command_queue_full",
  "session_pool_exhausted",
  "takeover_locked_out",
  "step_up_auth_required",
  "lease_expired",
  "stream_unavailable",
] as const;

export const liveBrowserApprovalDecisionValues = [
  "approved",
  "rejected",
] as const;

export const liveBrowserBarrierTypeValues = [
  "login_required",
  "captcha_required",
  "payment_review_required",
  "booking_confirmation_required",
] as const;

export const liveBrowserStreamScopeValues = [
  "viewer",
  "controller",
] as const;

export const liveBrowserActorTypeSchema = z.enum(liveBrowserActorTypeValues);
export const liveBrowserSourceTypeSchema = z.enum(liveBrowserSourceTypeValues);
export const liveBrowserSessionStatusSchema = z.enum(liveBrowserSessionStatusValues);
export const liveBrowserControlModeSchema = z.enum(liveBrowserControlModeValues);
export const liveBrowserAssistRequestTypeSchema = z.enum(liveBrowserAssistRequestTypeValues);
export const liveBrowserAssistRequestStatusSchema = z.enum(liveBrowserAssistRequestStatusValues);
export const liveBrowserEventTypeSchema = z.enum(liveBrowserEventTypeValues);
export const liveBrowserErrorCodeSchema = z.enum(liveBrowserErrorCodeValues);
export const liveBrowserApprovalDecisionSchema = z.enum(liveBrowserApprovalDecisionValues);
export const liveBrowserBarrierTypeSchema = z.enum(liveBrowserBarrierTypeValues);
export const liveBrowserStreamScopeSchema = z.enum(liveBrowserStreamScopeValues);

const stringRecordSchema = z.record(z.string(), z.unknown());

export const liveBrowserActorSchema = z.object({
  actorType: liveBrowserActorTypeSchema,
  actorId: z.string().min(1),
}).strict();

export const liveBrowserCommandSchema = z.object({
  type: z.literal("natural_language"),
  text: z.string().min(1),
}).strict();

export const liveBrowserAssistResponseSchema = z.discriminatedUnion("type", [
  z.object({
    type: z.literal("decision"),
    value: z.string().min(1),
  }).strict(),
  z.object({
    type: z.literal("field_input"),
    fields: stringRecordSchema,
  }).strict(),
  z.object({
    type: z.literal("review_page"),
    notes: z.string().min(1),
  }).strict(),
  z.object({
    type: z.literal("takeover_required"),
    reason: z.string().min(1),
  }).strict(),
]);

export const liveBrowserStreamSchema = z.object({
  viewerToken: z.string().min(1).optional(),
  controllerToken: z.string().min(1).optional(),
  expiresAt: z.string().min(1),
  leaseExpiresAt: z.string().min(1).optional(),
}).strict();

export const liveBrowserSessionSchema = z.object({
  sessionId: z.string().min(1),
  tenantId: z.string().min(1),
  userId: z.number().int().nonnegative(),
  sourceType: liveBrowserSourceTypeSchema,
  sourceId: z.string().min(1).nullable().optional(),
  status: liveBrowserSessionStatusSchema,
  controlMode: liveBrowserControlModeSchema,
  sessionVersion: z.number().int().nonnegative(),
  controllerActorType: liveBrowserActorTypeSchema.nullable().optional(),
  controllerActorId: z.string().min(1).nullable().optional(),
  controllerConnectionId: z.string().min(1).nullable().optional(),
  controllerLeaseExpiresAt: z.string().min(1).nullable().optional(),
  pauseReason: z.string().min(1).nullable().optional(),
  barrierType: liveBrowserBarrierTypeSchema.nullable().optional(),
  pendingAssistRequestId: z.string().min(1).nullable().optional(),
  pendingApprovalRequestId: z.string().min(1).nullable().optional(),
  policyContext: stringRecordSchema.default({}),
  browserContextRef: stringRecordSchema.default({}),
  stream: liveBrowserStreamSchema.optional(),
  activeTabCount: z.number().int().positive().default(1),
  startedAt: z.string().min(1),
  lastActivityAt: z.string().min(1),
  endedAt: z.string().min(1).nullable().optional(),
  endReason: z.string().min(1).nullable().optional(),
}).strict();

export const liveBrowserEventPayloadSchema = z.object({
  session: liveBrowserSessionSchema.optional(),
}).catchall(z.unknown());

export const liveBrowserEventEnvelopeSchema = z.object({
  eventId: z.string().min(1),
  sessionId: z.string().min(1),
  sessionVersion: z.number().int().nonnegative(),
  type: liveBrowserEventTypeSchema,
  timestamp: z.string().min(1),
  payload: liveBrowserEventPayloadSchema.default({}),
  cursor: z.string().min(1),
}).strict();

export const liveBrowserErrorSchema = z.object({
  code: liveBrowserErrorCodeSchema,
  message: z.string().min(1),
  currentSessionVersion: z.number().int().nonnegative().optional(),
  retryable: z.boolean(),
  reasonCodes: z.array(z.string()).default([]),
}).strict();

export const liveBrowserErrorResponseSchema = z.object({
  accepted: z.literal(false),
  error: liveBrowserErrorSchema,
}).strict();

const liveBrowserMutationRequestBaseSchema = z.object({
  sessionId: z.string().min(1),
  sessionVersion: z.number().int().nonnegative(),
  idempotencyKey: z.string().min(1),
  actor: liveBrowserActorSchema,
}).strict();

export const liveBrowserCreateSessionRequestSchema = z.object({
  actor: liveBrowserActorSchema,
  sourceType: liveBrowserSourceTypeSchema,
  sourceId: z.string().min(1).nullable().optional(),
  initialUrl: z.string().min(1).optional(),
  mode: liveBrowserControlModeSchema.default("observe"),
  executionIntent: z.object({
    prompt: z.string().min(1),
    skillId: browserSkillIdSchema.optional(),
    discoverWebsites: z.boolean().optional(),
    autoDraftSkill: z.boolean().optional(),
  }).strict().optional(),
}).strict();

export const liveBrowserCreateSessionResponseSchema = z.object({
  sessionId: z.string().min(1),
  status: liveBrowserSessionStatusSchema,
  controlMode: liveBrowserControlModeSchema,
  sessionVersion: z.number().int().nonnegative(),
  stream: liveBrowserStreamSchema,
}).strict();

export const liveBrowserGetSessionRequestSchema = z.object({
  sessionId: z.string().min(1),
  actor: liveBrowserActorSchema,
}).strict();

export const liveBrowserSendCommandRequestSchema = liveBrowserMutationRequestBaseSchema.extend({
  command: liveBrowserCommandSchema,
}).strict();

export const liveBrowserSendCommandResponseSchema = z.object({
  accepted: z.literal(true),
  sessionVersion: z.number().int().nonnegative(),
  queuedCommandId: z.string().min(1),
}).strict();

export const liveBrowserUpdatePolicyContextRequestSchema = liveBrowserMutationRequestBaseSchema.extend({
  policyContextPatch: stringRecordSchema.default({}),
}).strict();

export const liveBrowserUpdatePolicyContextResponseSchema = z.object({
  accepted: z.literal(true),
  sessionVersion: z.number().int().nonnegative(),
  policyContext: stringRecordSchema.default({}),
}).strict();

export const liveBrowserPauseAgentRequestSchema = liveBrowserMutationRequestBaseSchema.extend({
  reason: z.string().min(1),
}).strict();

export const liveBrowserPauseAgentResponseSchema = z.object({
  accepted: z.literal(true),
  status: liveBrowserSessionStatusSchema,
  controlMode: liveBrowserControlModeSchema,
  sessionVersion: z.number().int().nonnegative(),
}).strict();

export const liveBrowserTakeControlRequestSchema = liveBrowserMutationRequestBaseSchema.extend({
  reason: z.string().min(1),
  stepUpCode: z.string().min(1).optional(),
}).strict();

export const liveBrowserTakeControlResponseSchema = z.object({
  accepted: z.literal(true),
  status: liveBrowserSessionStatusSchema,
  controlMode: liveBrowserControlModeSchema,
  sessionVersion: z.number().int().nonnegative(),
  stream: liveBrowserStreamSchema,
}).strict();

export const liveBrowserReturnControlRequestSchema = liveBrowserMutationRequestBaseSchema.extend({
  checkpoint: z.string().min(1),
  notes: z.string().min(1).optional(),
}).strict();

export const liveBrowserReturnControlResponseSchema = z.object({
  accepted: z.literal(true),
  status: liveBrowserSessionStatusSchema,
  controlMode: liveBrowserControlModeSchema,
  sessionVersion: z.number().int().nonnegative(),
}).strict();

export const liveBrowserSubmitAssistResponseRequestSchema = liveBrowserMutationRequestBaseSchema.extend({
  assistRequestId: z.string().min(1),
  response: liveBrowserAssistResponseSchema,
}).strict();

export const liveBrowserSubmitAssistResponseResponseSchema = z.object({
  accepted: z.literal(true),
  assistRequestStatus: liveBrowserAssistRequestStatusSchema,
  sessionVersion: z.number().int().nonnegative(),
}).strict();

export const liveBrowserResolveApprovalRequestSchema = liveBrowserMutationRequestBaseSchema.extend({
  approvalRequestId: z.string().min(1),
  decision: liveBrowserApprovalDecisionSchema,
  notes: z.string().min(1).optional(),
}).strict();

export const liveBrowserResolveApprovalResponseSchema = z.object({
  accepted: z.literal(true),
  approvalStatus: liveBrowserApprovalDecisionSchema,
  sessionVersion: z.number().int().nonnegative(),
  agentResumed: z.boolean(),
}).strict();

export const liveBrowserCancelSessionRequestSchema = liveBrowserMutationRequestBaseSchema.extend({
  reason: z.string().min(1),
}).strict();

export const liveBrowserCancelSessionResponseSchema = z.object({
  accepted: z.literal(true),
  status: liveBrowserSessionStatusSchema,
  sessionVersion: z.number().int().nonnegative(),
}).strict();

export const liveBrowserListEventsRequestSchema = z.object({
  sessionId: z.string().min(1),
  actor: liveBrowserActorSchema,
  cursor: z.string().min(1).optional(),
  limit: z.number().int().positive().max(500).default(100),
}).strict();

export const liveBrowserListEventsResponseSchema = z.object({
  sessionId: z.string().min(1),
  events: z.array(liveBrowserEventEnvelopeSchema),
  nextCursor: z.string().min(1).nullable().optional(),
  hasMore: z.boolean(),
}).strict();

export const liveBrowserStreamTokenRequestSchema = z.object({
  sessionId: z.string().min(1),
  actor: liveBrowserActorSchema,
  scope: liveBrowserStreamScopeSchema,
}).strict();

export const liveBrowserStreamTokenResponseSchema = z.object({
  sessionId: z.string().min(1),
  scope: liveBrowserStreamScopeSchema,
  token: z.string().min(1),
  expiresAt: z.string().min(1),
  leaseExpiresAt: z.string().min(1).nullable().optional(),
}).strict();

export type LiveBrowserActor = z.infer<typeof liveBrowserActorSchema>;
export type LiveBrowserSession = z.infer<typeof liveBrowserSessionSchema>;
export type LiveBrowserEventPayload = z.infer<typeof liveBrowserEventPayloadSchema>;
export type LiveBrowserEventEnvelope = z.infer<typeof liveBrowserEventEnvelopeSchema>;
export type LiveBrowserError = z.infer<typeof liveBrowserErrorSchema>;
export type LiveBrowserCreateSessionRequest = z.infer<typeof liveBrowserCreateSessionRequestSchema>;
export type LiveBrowserCreateSessionResponse = z.infer<typeof liveBrowserCreateSessionResponseSchema>;
export type LiveBrowserGetSessionRequest = z.infer<typeof liveBrowserGetSessionRequestSchema>;
export type LiveBrowserSendCommandRequest = z.infer<typeof liveBrowserSendCommandRequestSchema>;
export type LiveBrowserSendCommandResponse = z.infer<typeof liveBrowserSendCommandResponseSchema>;
export type LiveBrowserUpdatePolicyContextRequest = z.infer<typeof liveBrowserUpdatePolicyContextRequestSchema>;
export type LiveBrowserUpdatePolicyContextResponse = z.infer<typeof liveBrowserUpdatePolicyContextResponseSchema>;
export type LiveBrowserPauseAgentRequest = z.infer<typeof liveBrowserPauseAgentRequestSchema>;
export type LiveBrowserPauseAgentResponse = z.infer<typeof liveBrowserPauseAgentResponseSchema>;
export type LiveBrowserTakeControlRequest = z.infer<typeof liveBrowserTakeControlRequestSchema>;
export type LiveBrowserTakeControlResponse = z.infer<typeof liveBrowserTakeControlResponseSchema>;
export type LiveBrowserReturnControlRequest = z.infer<typeof liveBrowserReturnControlRequestSchema>;
export type LiveBrowserReturnControlResponse = z.infer<typeof liveBrowserReturnControlResponseSchema>;
export type LiveBrowserSubmitAssistResponseRequest = z.infer<typeof liveBrowserSubmitAssistResponseRequestSchema>;
export type LiveBrowserSubmitAssistResponseResponse = z.infer<typeof liveBrowserSubmitAssistResponseResponseSchema>;
export type LiveBrowserResolveApprovalRequest = z.infer<typeof liveBrowserResolveApprovalRequestSchema>;
export type LiveBrowserResolveApprovalResponse = z.infer<typeof liveBrowserResolveApprovalResponseSchema>;
export type LiveBrowserCancelSessionRequest = z.infer<typeof liveBrowserCancelSessionRequestSchema>;
export type LiveBrowserCancelSessionResponse = z.infer<typeof liveBrowserCancelSessionResponseSchema>;
export type LiveBrowserListEventsRequest = z.infer<typeof liveBrowserListEventsRequestSchema>;
export type LiveBrowserListEventsResponse = z.infer<typeof liveBrowserListEventsResponseSchema>;
export type LiveBrowserStreamTokenRequest = z.infer<typeof liveBrowserStreamTokenRequestSchema>;
export type LiveBrowserStreamTokenResponse = z.infer<typeof liveBrowserStreamTokenResponseSchema>;

export function getLiveBrowserEventSessionSnapshot(
  event: Pick<LiveBrowserEventEnvelope, "payload">,
): LiveBrowserSession | null {
  const candidate = event.payload?.session;
  if (!candidate) {
    return null;
  }
  const parsed = liveBrowserSessionSchema.safeParse(candidate);
  return parsed.success ? parsed.data : null;
}

File: python-backend/app/models/live_browser.py
"""Durable live-browser runtime models."""

from __future__ import annotations

from datetime import datetime
import enum

from sqlalchemy import JSON, Column, DateTime, ForeignKey, Index, Integer, String, Text
from sqlalchemy.orm import relationship

from app.core.database import Base


class LiveBrowserSourceType(str, enum.Enum):
    AUTOMATION = "automation"
    WORKFLOW = "workflow"
    AGENCY = "agency"


class LiveBrowserSessionStatus(str, enum.Enum):
    CREATED = "created"
    PROVISIONING = "provisioning"
    READY = "ready"
    AGENT_RUNNING = "agent_running"
    WAITING_FOR_HUMAN = "waiting_for_human"
    HUMAN_CONTROLLING = "human_controlling"
    WAITING_FOR_RUNTIME_RECOVERY = "waiting_for_runtime_recovery"
    FAILED_RECOVERY_REQUIRED = "failed_recovery_required"
    COMPLETED = "completed"
    CANCELLED = "cancelled"
    FAILED = "failed"
    EXPIRED = "expired"


class LiveBrowserControlMode(str, enum.Enum):
    OBSERVE = "observe"
    APPROVE_ONLY = "approve_only"
    TAKEOVER = "takeover"
    AGENT_CONTROL = "agent_control"


class LiveBrowserActorType(str, enum.Enum):
    AGENT = "agent"
    USER = "user"
    SYSTEM = "system"
    POLICY = "policy"


class LiveBrowserEventType(str, enum.Enum):
    SESSION_CREATED = "session_created"
    SESSION_STATE_CHANGED = "session_state_changed"
    STREAM_READY = "stream_ready"
    FRAME_UPDATED = "frame_updated"
    URL_CHANGED = "url_changed"
    COMMAND_QUEUED = "command_queued"
    COMMAND_STARTED = "command_started"
    COMMAND_COMPLETED = "command_completed"
    COMMAND_FAILED = "command_failed"
    ASSIST_REQUESTED = "assist_requested"
    ASSIST_RESOLVED = "assist_resolved"
    APPROVAL_REQUESTED = "approval_requested"
    APPROVAL_RESOLVED = "approval_resolved"
    TAKEOVER_STARTED = "takeover_started"
    TAKEOVER_LEASE_EXPIRING = "takeover_lease_expiring"
    TAKEOVER_ENDED = "takeover_ended"
    INCIDENT = "incident"
    AGENT_STARTED = "agent_started"
    AGENT_RESUMED = "agent_resumed"
    NAVIGATION_COMPLETED = "navigation_completed"
    SESSION_COMPLETED = "session_completed"
    SESSION_FAILED = "session_failed"


class LiveBrowserSession(Base):
    __tablename__ = "live_browser_sessions"

    id = Column(String(64), primary_key=True)
    tenant_id = Column("tenantId", String(36), ForeignKey("tenants.id", ondelete="CASCADE"), nullable=False)
    user_id = Column("userId", Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    source_type = Column("sourceType", String(32), nullable=False)
    source_id = Column("sourceId", String(128), nullable=True)
    status = Column("status", String(64), nullable=False, default=LiveBrowserSessionStatus.CREATED.value)
    control_mode = Column("controlMode", String(64), nullable=False, default=LiveBrowserControlMode.OBSERVE.value)
    session_version = Column("sessionVersion", Integer, nullable=False, default=1)
    controller_actor_type = Column("controllerActorType", String(32), nullable=True)
    controller_actor_id = Column("controllerActorId", String(64), nullable=True)
    controller_connection_id = Column("controllerConnectionId", String(128), nullable=True)
    controller_lease_expires_at = Column("controllerLeaseExpiresAt", DateTime(timezone=True), nullable=True)
    runtime_owner_id = Column("runtimeOwnerId", String(128), nullable=True)
    runtime_owner_claimed_at = Column("runtimeOwnerClaimedAt", DateTime(timezone=True), nullable=True)
    pause_reason = Column("pauseReason", String(128), nullable=True)
    pending_assist_request_id = Column("pendingAssistRequestId", String(64), nullable=True)
    pending_approval_request_id = Column("pendingApprovalRequestId", String(64), nullable=True)
    policy_context_json = Column("policyContextJson", JSON, nullable=False, default=dict)
    browser_context_ref = Column("browserContextRef", JSON, nullable=False, default=dict)
    stream_ref = Column("streamRef", JSON, nullable=False, default=dict)
    active_tab_count = Column("activeTabCount", Integer, nullable=False, default=1)
    started_at = Column("startedAt", DateTime(timezone=True), nullable=False, default=datetime.utcnow)
    last_activity_at = Column("lastActivityAt", DateTime(timezone=True), nullable=False, default=datetime.utcnow)
    ended_at = Column("endedAt", DateTime(timezone=True), nullable=True)
    end_reason = Column("endReason", String(128), nullable=True)

    events = relationship("LiveBrowserEvent", back_populates="session", cascade="all, delete-orphan")
    idempotency_keys = relationship(
        "LiveBrowserIdempotencyKey",
        back_populates="session",
        cascade="all, delete-orphan",
    )

    __table_args__ = (
        Index("live_browser_sessions_tenant_status_idx", "tenantId", "status"),
        Index("live_browser_sessions_user_activity_idx", "userId", "lastActivityAt"),
        Index("live_browser_sessions_runtime_owner_idx", "runtimeOwnerId", "runtimeOwnerClaimedAt"),
    )


class LiveBrowserIdempotencyKey(Base):
    __tablename__ = "live_browser_idempotency_keys"

    id = Column(Integer, primary_key=True, autoincrement=True)
    session_id = Column("sessionId", String(64), ForeignKey("live_browser_sessions.id", ondelete="CASCADE"), nullable=False)
    idempotency_key = Column("idempotencyKey", String(128), nullable=False)
    command_type = Column("commandType", String(64), nullable=False)
    response_json = Column("responseJson", JSON, nullable=False, default=dict)
    created_at = Column("createdAt", DateTime(timezone=True), nullable=False, default=datetime.utcnow)
    expires_at = Column("expiresAt", DateTime(timezone=True), nullable=False)

    session = relationship("LiveBrowserSession", back_populates="idempotency_keys")

    __table_args__ = (
        Index("uq_live_browser_idempotency_keys_session_key", "sessionId", "idempotencyKey", unique=True),
        Index("live_browser_idempotency_keys_expires_idx", "expiresAt"),
    )


class LiveBrowserEvent(Base):
    __tablename__ = "live_browser_events"

    id = Column(String(64), primary_key=True)
    session_id = Column("sessionId", String(64), ForeignKey("live_browser_sessions.id", ondelete="CASCADE"), nullable=False)
    session_version_at = Column("sessionVersionAt", Integer, nullable=False)
    tenant_id = Column("tenantId", String(36), ForeignKey("tenants.id", ondelete="CASCADE"), nullable=False)
    event_type = Column("eventType", String(64), nullable=False)
    actor_type = Column("actorType", String(32), nullable=False)
    actor_id = Column("actorId", String(64), nullable=True)
    payload_json = Column("payloadJson", JSON, nullable=False, default=dict)
    screenshot_ref = Column("screenshotRef", String(255), nullable=True)
    cursor = Column("cursor", String(255), nullable=False)
    created_at = Column("createdAt", DateTime(timezone=True), nullable=False, default=datetime.utcnow)

    session = relationship("LiveBrowserSession", back_populates="events")

    __table_args__ = (
        Index("uq_live_browser_events_session_cursor", "sessionId", "cursor", unique=True),
        Index("live_browser_events_session_created_idx", "sessionId", "createdAt"),
        Index("live_browser_events_session_version_idx", "sessionId", "sessionVersionAt"),
    )


class LiveBrowserAssistRequest(Base):
    __tablename__ = "live_browser_assist_requests"

    id = Column(String(64), primary_key=True)
    session_id = Column("sessionId", String(64), ForeignKey("live_browser_sessions.id", ondelete="CASCADE"), nullable=False)
    session_version_at = Column("sessionVersionAt", Integer, nullable=False)
    tenant_id = Column("tenantId", String(36), ForeignKey("tenants.id", ondelete="CASCADE"), nullable=False)
    request_type = Column("requestType", String(32), nullable=False)
    status = Column(String(32), nullable=False, default="pending")
    prompt = Column(Text, nullable=False)
    context_json = Column("contextJson", JSON, nullable=False, default=dict)
    response_json = Column("responseJson", JSON, nullable=False, default=dict)
    resolved_session_version_at = Column("resolvedSessionVersionAt", Integer, nullable=True)
    requested_at = Column("requestedAt", DateTime(timezone=True), nullable=False, default=datetime.utcnow)
    resolved_at = Column("resolvedAt", DateTime(timezone=True), nullable=True)

    __table_args__ = (
        Index("live_browser_assist_requests_session_status_idx", "sessionId", "status"),
        Index("live_browser_assist_requests_session_requested_idx", "sessionId", "requestedAt"),
    )


class LiveBrowserControlTransfer(Base):
    __tablename__ = "live_browser_control_transfers"

    id = Column(String(64), primary_key=True)
    session_id = Column("sessionId", String(64), ForeignKey("live_browser_sessions.id", ondelete="CASCADE"), nullable=False)
    session_version_at = Column("sessionVersionAt", Integer, nullable=False)
    tenant_id = Column("tenantId", String(36), ForeignKey("tenants.id", ondelete="CASCADE"), nullable=False)
    from_actor_type = Column("fromActorType", String(32), nullable=False)
    from_actor_id = Column("fromActorId", String(64), nullable=True)
    to_actor_type = Column("toActorType", String(32), nullable=False)
    to_actor_id = Column("toActorId", String(64), nullable=True)
    reason = Column(String(128), nullable=False)
    policy_check_hash = Column("policyCheckHash", String(128), nullable=True)
    created_at = Column("createdAt", DateTime(timezone=True), nullable=False, default=datetime.utcnow)

    __table_args__ = (
        Index("live_browser_control_transfers_session_created_idx", "sessionId", "createdAt"),
    )
